
<a href="https://colab.research.google.com/github/aleylani/Databehandling-AI24/blob/main/exercises/E3_merging.ipynb" target="_parent"><img align="left" src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a> &nbsp; to see hints and answers.

---
# Merging exercises

---
These are introductory exercises in Pandas with focus in **syntax, indexing, data selection, missing data, aggregations, visualizations**, **data cleaning**, **merging**, **concatenation**, **joining**, **parse html tables** .

<p class = "alert alert-info" role="alert"><b>Note</b> all datasets used in this exercise can be found under Data folder of the course Github repo</p>

<p class = "alert alert-info" role="alert"><b>Note</b> that there are usually many ways to achieve same results using operations in pandas. Feel free to investigate several ways. </p>

<p class = "alert alert-info" role="alert"><b>Note</b> that in cases when you start to repeat code, try not to. Create functions to reuse code instead. </p>

<p class = "alert alert-info" role="alert"><b>Remember</b> to use <b>descriptive variable, function, index </b> and <b> column names</b> in order to get readable code </p>

<p class = "alert alert-info" role="alert"><b>Remember</b> to format your input questions in a pedagogical way to guide the user

The number of stars (\*), (\*\*), (\*\*\*) denotes the difficulty level of the task

---

**Note that you might need to install lxml in your environment**

In [26]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

----
**A)**

In [28]:
se_demo_df = pd.read_html('https://sv.wikipedia.org/wiki/Sveriges_demografi')



---
**B)**

In [29]:
df_1b = se_demo_df[6]
df_1b
sns.histplot(data=df_1b, x="Folkmängd", kde=True, bins=30)

#plt.show()



ValueError: Could not interpret value `Folkmängd` for `x`. An entry with this name does not appear in `data`.

----
**C)**

In [32]:
df_1c = se_demo_df[1]
df_1c2 = pd.DataFrame()
df_1c2["År"] = df_1c["Vid utgången av år"][0:7]
df_1c2["Folkmängd"] = df_1c["Folkmängd"][0:7]

df_1c2 = df_1c2.set_index("År")
df_1c2

Folkmängd
År             
1570    900 000
1650  1 225 000
1700  1 485 000
1720  1 350 000
1755  1 878 000
1815  2 465 000
1865  4 099 000

---
**D)**

In [47]:
df_1c2
#df_1b_clean = df_1b.set_index("År")
#df_1d = df_1b_clean.merge(df_1c2_clean,on="År")
#df_1d

Folkmängd
År             
1570    900 000
1650  1 225 000
1700  1 485 000
1720  1 350 000
1755  1 878 000
1815  2 465 000
1865  4 099 000

In [8]:

df_1b
#print(df_1c2.head())

df_1b_clean = df_1b.drop(["Födda", "Döda", "Naturlig befolknings- förändring", "Nativitet (per 1000)", "Dödstal (per 1000)", "Naturlig befolknings- förändring (per 1000)", "Total fertilitet"], axis=1)
df_1b_clean = df_1b_clean.rename(columns={"Unnamed: 0": "År", "Medel- folkmängd": "Folkmängd"})
print(df_1b_clean)


df_1c2_clean = df_1c2
df_1c2_clean["Folkmängd"] = df_1c2["Folkmängd"].astype(float)
print(df_1c2)


df_1d = df_1b_clean.merge(df_1c2,on="År")
df_1d

       År     Folkmängd
0    1900     5 117 000
1    1901     5 156 000
2    1902     5 187 000
3    1903     5 210 000
4    1904     5 241 000
..    ...           ...
118  2018  10 175 213,5
119  2019  10 278 887,0
120  2020  10 353 442,0
121  2021  10 415 810,5
122  2022  10 521 556,0

[123 rows x 2 columns]


ValueError: could not convert string to float: '900\xa0000'

## 1. Swedish demographic data (*)

Go to Swedish-language wikipedia page [Sveriges demografi](https://sv.wikipedia.org/wiki/Sveriges_demografi). 

&nbsp; a) Read in the table under "Befolkningsstatistik sedan 1900" into a DataFrame

          Note that you might need to install lxml in your environment

          conda install lxml

&nbsp; b) Choose to do some EDA (exploratory data analysis) on this dataset. And draw some relevant graphs.

&nbsp; c) Now we want to go backwards in time (before 1900) to see how population has changed in Sweden. Read in the table under history and keep the data of "Folkmängd" from 1570-1865. 

| År   | Folkmängd |
| ---- | --------- | 
| 1570 | 900000    |     
| 1650 | 1225000   |
| 1700 | 1485000   |
| 1720 | 1350000   |
| 1755 | 1878000   |
| 1815 | 2465000   |
| 1865 | 4099000   |


&nbsp; d) Now concatenate this with the table from 1900 so that you have population data from 1570 to 2020. Note that you may need to clean the data in order for it to fit properly. Also you may be able to do this in several ways. 

&nbsp; e) Draw a graph of population data from 1570-2020.

&nbsp; f) Try to see if you can plot something from this data that might show the effects of the Covid-19 pandemic. It's good to zoom in the data, i.e. plot a shorter time frame instead of all time frames. 

&nbsp; g) Try to see if you can plot something from this data that might capture the effects imigration and emigration waves.

&nbsp; h) Feel free to find other interesting facts about the Swedish demographics based on this data. 

<details>
<summary>Hint</summary>

Useful methods:
- append()
- join()
- concat()
- merge()

</details>

<br/>

<details>

<summary>Answer</summary>

![Sweden population data 1952-2020](../assets/sverige_befolkning_tid.png)

</details>

## 1. Swedish demographic data (*)

Go to Swedish-language wikipedia page [Sveriges demografi](https://sv.wikipedia.org/wiki/Sveriges_demografi). 

&nbsp; a) Read in the table under "Befolkningsstatistik sedan 1900" into a DataFrame

          Note that you might need to install lxml in your environment

          conda install lxml

&nbsp; b) Choose to do some EDA (exploratory data analysis) on this dataset. And draw some relevant graphs.

&nbsp; c) Now we want to go backwards in time (before 1900) to see how population has changed in Sweden. Read in the table under history and keep the data of "Folkmängd" from 1570-1865. 

| År   | Folkmängd |
| ---- | --------- | 
| 1570 | 900000    |     
| 1650 | 1225000   |
| 1700 | 1485000   |
| 1720 | 1350000   |
| 1755 | 1878000   |
| 1815 | 2465000   |
| 1865 | 4099000   |


&nbsp; d) Now concatenate this with the table from 1900 so that you have population data from 1570 to 2020. Note that you may need to clean the data in order for it to fit properly. Also you may be able to do this in several ways. 

&nbsp; e) Draw a graph of population data from 1570-2020.

&nbsp; f) Try to see if you can plot something from this data that might show the effects of the Covid-19 pandemic. It's good to zoom in the data, i.e. plot a shorter time frame instead of all time frames. 

&nbsp; g) Try to see if you can plot something from this data that might capture the effects imigration and emigration waves.

&nbsp; h) Feel free to find other interesting facts about the Swedish demographics based on this data. 

<details>
<summary>Hint</summary>

Useful methods:
- append()
- join()
- concat()
- merge()

</details>

<br/>

<details>

<summary>Answer</summary>

![Sweden population data 1952-2020](../assets/sverige_befolkning_tid.png)

</details>

---
## 2. Denmark demographic data (*)

Go to the Danish-language wikipedia page [Danmarks demografi](https://da.wikipedia.org/wiki/Danmarks_demografi). 

&nbsp; a) Read in the table under "Demografiske data" into a DataFrame (*)

&nbsp; b) Clean the data and draw a graph of population against year from 1769-2020. (**)


In [22]:
import pandas as pd

dk_demo_df = pd.read_html('https://da.wikipedia.org/wiki/Danmarks_demografi')[0]


dk_år_filter_1779_1975 = dk_demo_df[0]['År'] >= 1769


type(dk_demo_df)

#dk_demo_df[0][dk_år_filter]
#dk_demo_df.tail()

KeyError: 0

---
## 3. Norwegian demographic data (*)

Go to Swedish-language wikipedia page [Norges demografi](https://sv.wikipedia.org/wiki/Norges_demografi). 

&nbsp; a) Read in the table under "Befolkningsstatistik sedan 1900" into a DataFrame

&nbsp; b) You see some missing data in column "Total fertilitet". Go to the [English page](https://en.wikipedia.org/wiki/Demographics_of_Norway) and read in the data from "Vital statistics since 1900".  

&nbsp; c) Pick out the fertility column from b) dataset, merge it into a) dataset and clean the data so that you only have columns "År", "Folkmängd", "Fertilitet". 


---
## 4. Merge Sweden-Norway (*)

Create a population graph and a fertility graph showing Sweden and Norway.

<details>

<summary>Answer</summary>

![Fertilitet Norge och Sverige](../assets/fertilitet_sv_no.png)

![Folkmängd Norge och Sverige](../assets/folkmangd_sverige_norge.png)

</details>